File: segment_tree.py
import sys
import math
from collections import defaultdict,Counter

input=sys.stdin.readline

def print_(x):
    sys.stdout.write(str(x)+" ")
def rInt():
    return int(input())
def rList():
    arr = input().split()
    return arr
def cInt(arr):
    res = []
    for itr in arr:
        res.append(int(itr))
    return res
def pList(arr):
    for _ in arr:
        print_(_)
def iniList(size,ini_val):
    arr = [ ini_val for _ in range(size)]
    return arr

# point add update
# range add update
# point sum query
# range sum query
# All are O(logn)

class Segment_Tree:
    tree=[]# 0 indexed
    lazy=[]
    vals=[]
    n = 0
    def __init__(self,arr=[]):
        self.n = len(arr)
        self.vals = arr
        self.tree = iniList(4*self.n,0)
        self.lazy = iniList(4*self.n,0)
        
    def inputVals(self,arr):
        self.vals=arr
        self.n = len(arr)
        self.tree = iniList(4*self.n,0)
        self.lazy = iniList(4*self.n,0)
        
    def buildSeg(self,node,lft,rht):
        if lft==rht:
            self.tree[node]=self.vals[lft]
            return
        else:
            mid = lft + (rht-lft)//2
            self.buildSeg(node*2+1,lft,mid)
            self.buildSeg(node*2+2,mid+1,rht)
            self.tree[node]=self.tree[(node*2)+1]+self.lazy[(node*2)+1]*(mid-lft+1)
            self.tree[node]+=self.tree[(node*2)+2]+self.lazy[(node*2)+2]*(rht-mid)
        
    def build(self):
        self.buildSeg(0,0,self.n-1)
    
    def upd(self,node,lft,rht,L,R,new_val):
        if rht<L or lft>R:
            return
        elif L<=lft and rht<=R:
            self.lazy[node]+=new_val
        else:
            mid = lft + (rht-lft)//2
            if lft<=R and mid>=L: 
                self.upd((node*2)+1,lft,mid,L,R,new_val)
            if mid+1<=R and rht>=L: 
                self.upd((node*2)+2,mid+1,rht,L,R,new_val)
            self.tree[node]=self.tree[(node*2)+1]+self.lazy[(node*2)+1]*(mid-lft+1)
            self.tree[node]+=self.tree[(node*2)+2]+self.lazy[(node*2)+2]*(rht-mid)
    
    def updatePoi(self,idx,new_val):
        self.upd(0,0,self.n-1,idx,idx,new_val)
    def updateRng(self,L,R,new_val):
        self.upd(0,0,self.n-1,L,R,new_val)
    
    def push(self,node):
        tochild = self.lazy[node]
        self.lazy[node]=0
        self.lazy[node*2+1]+=tochild
        self.lazy[node*2+2]+=tochild
    
    def qry(self,node,lft,rht,L,R):
        if rht<L or lft>R:
            return 0
        elif L<=lft and rht<=R:
            return (self.tree[node]+(self.lazy[node]*(rht-lft+1)))
        res = 0
        self.push(node)
        mid = lft + (rht-lft)//2
        if mid<L or lft>R:
            res+=0
        else:  
            res = self.qry(node*2+1,lft,mid,L,R)
        if rht<L or mid+1>R:
            res+=0
        else:  
            res += self.qry(node*2+2,mid+1,rht,L,R)
        self.tree[node]=self.tree[(node*2)+1]+self.lazy[(node*2)+1]*(mid-lft+1)
        self.tree[node]+=self.tree[(node*2)+2]+self.lazy[(node*2)+2]*(rht-mid)
        return res
    
    def queryPoi(self,idx):
        return self.qry(0,0,self.n-1,idx,idx)
    def queryRng(self,L,R):
        return self.qry(0,0,self.n-1,L,R)

test_cases=rInt()
for test_case in range(test_cases):
    inp = rList()
    inp = cInt(inp)
    n = inp[0]
    q = inp[1]
    
    arr = iniList(n,0)
    seg = Segment_Tree(arr)

    ans = []
    for qq in range(q):
        rus = rList()
        rus = cInt(rus)
        if rus[0]==1:
            ans.append(seg.queryRng(rus[1]-1,rus[2]-1))
        else:
            seg.updateRng(rus[1]-1,rus[2]-1,rus[3])
    for _ in ans:
        print(_)
    

File: c++/segment_tree.cpp
#include <bits/stdc++.h>
using namespace std;

#define ll long long int 
int tree[2000005],a[100005];

void build(int node, int start, int end){
    if(start == end){
        // Leaf node will have a single element
        tree[node] = a[start];
    }
    else{
        int mid = (start + end) / 2;
        build(2*node, start, mid);
        build(2*node+1, mid+1, end);
        tree[node] = tree[2*node] + tree[2*node+1];
    }
}

void update(int node,int start,int end,int idx,int val){
    
	if(start==end){
		a[idx]=val;
		tree[node]=val;
	}
	else{
    	int mid=(start+end)/2;
    	
    	if(idx>=start&&idx<=mid)
    		update(2*node,start,mid,idx,val);
    	else
    		update(2*node+1,mid+1,end,idx,val);
    
    	if(tree[2*node]<tree[2*node+1])
    		tree[node]=tree[2*node];
    	else
    		tree[node]=tree[2*node+1];
	}
}
int query(int node, int start, int end, int l, int r)
{
    if(r < start or end < l){
        return 0;
    }
    if(l <= start and end <= r){
        cout<<start<<" "<<end<<" "<<tree[node]<<endl;
        return tree[node];
    }
    int mid = (start + end) / 2;
    int p1 = query(2*node, start, mid, l, r);
    int p2 = query(2*node+1, mid+1, end, l, r);
    return (p1 + p2);
}
int main(){

	int n,q;
	cin>>n>>q;
	
	for(int i=1;i<=n;i++)
		cin>>a[i];
		
	build(1,1,n);

	while(q--){
    	char b;
    	int x,y;
    	cin>>b>>x>>y;
    	if(b=='q'){
    		int c=query(1,1,n,x,y);
    		cout<<c<<endl;
    	}
    	else
    		update(1,1,n,x,y);
	}
	return 0;
}